README.md
# SKAlert

Alerts in iOS might be painful when it comes to writing multiple lines of code and has reusabilty issues. For that matter we have created `SKAlert` for simpler usage of alerts for iOS.

![](https://user-images.githubusercontent.com/40866243/42735853-4e0f32f2-8879-11e8-8fb4-1ebf14987998.gif)


## Requirements

`SKAlert` works for iOS 9.0+ and requires ARC to buid.

## Add SKAlert to your project

### Cocoapods ( Highly recommended way is to use cocoapods)
`pod 'SKAlert'`

### Static library

Follow this article at [Medium](https://medium.com/@nishantnitb/writing-custom-universal-framework-in-xcode-9-and-ios-11-7a63a2ce024a). But before continuing with the above article, delete Example project.

### Usage:

1. import SKAlert
2. Start Using It, by below sytax
```swift
_  = SKAlert().showAlert("Plain Alert Header.", subTitle: "Plain Alert Sub Title")
```
```swift
_  = SKAlert().showAlertWithOkAction("Alert Header", subTitle: "Alert Sub Title", okCompletionHandler: {
print("Ok Pressed")
})
```
```swift
_  = SKAlert().showAlertWithTwoButtons("Alert Header With Two Buttons", subTitle: "Alert Sub Title With Two Buttons", okCompletionHandler: {
print("Ok Pressed")
}, cancelCompletionHandler: {
print("Canel Pressed")
})
```
```swift
_ = SKAlert().showAlertWithCustomButtons("Alert Header With Two Custom Buttons", subTitle: "Sub title with custom buttons with Handlers........!!!", leftBtnTitle: "Left", leftBtnColor: UIColor.black, rightBtnTitle: "right", rightBtnColor: UIColor.red, leftCompletionHandler: {
print("left")
}) {
print("right")
}
```
```swift
let alert = SKAlert()
alert.isTapToDismiss = false
alert.showAlert("Plain Alert Header.", subTitle: "Plain Alert Sub Title")
```

### Noticeable mentions

1. SweetAlert
2. Medium [article](https://medium.com/flawless-app-stories/getting-started-with-reusable-frameworks-for-ios-development-f00d74827d11)
## License

This code is distributed under the terms and conditions of the [MIT license](LICENSE).

## Change-log

A brief summary of each release can be found in the [CHANGELOG](CHANGELOG.md). 


CHANGELOG.md
**Version 1.1.2** @ 15.07.2018

- If you pass title/ subtitle empty to SKAlert fucntion then that field automatically hides
- For alerts containig completion handlers. You will have to manaully set 'isTapToDismiss' variable to false, to disable tap anywhere to dismiss feature. 
- Updated Animations.


SKAlert/SKAlert.swift
//
//  SKAlert.swift
//  SKAlert
//
//  Created by Srujan k on 04/07/18.
//  Copyright © 2018 Srujan k. All rights reserved.
//

import UIKit

public typealias OkCompletionHandler = () -> Void
public typealias CancelCompletionHandler = () -> Void


public class SKAlert: UIViewController {
    @IBOutlet private weak var headerTitle: UILabel!
    @IBOutlet private weak var subTitle: UILabel!
    @IBOutlet private weak var cancelButton: UIButton!
    @IBOutlet private weak var okButton: UIButton!
    @IBOutlet private weak var mainView: UIView!
    @IBOutlet private weak var subView: UIView!
    
    var strongSelf: SKAlert?
    let SKCornerRadius: CGFloat = 3.0
    var contentView =  UIView()
    let nibname = "SKAlert"
    var OkayButtonTag = 1
    let CancelButtonTag = 2
    var okCompletion: OkCompletionHandler? = nil
    var cancelCompletion: CancelCompletionHandler? =  nil

    public init() {
        super.init(nibName: nil, bundle: nil)
        strongSelf = self
        self.setUp()
    }
    // MARK: - Set up
    func setUp() {
        let bundle = Bundle(for: type(of: self))
        let nib = UINib(nibName: self.nibname, bundle: bundle)
        self.view = nib.instantiate(withOwner: self, options: nil).first as! UIView
        self.view.frame = UIScreen.main.bounds
    }
    func setUpContentView() {
        strongSelf?.contentView.layer.cornerRadius = 5.0
        strongSelf?.contentView.layer.borderWidth = 2
        self.okButton.layer.cornerRadius = SKCornerRadius
        self.cancelButton.layer.cornerRadius = SKCornerRadius
        self.contentView.layer.masksToBounds = true

    }
    required public init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    public func showAlert(_ title: String, subTitle: String){
        let window: UIWindow = UIApplication.shared.keyWindow!
        window.addSubview(view)
        window.bringSubview(toFront: view)
        view.frame = window.bounds
        self.setUpContentView()
//        view.addSubview(self.subView)
        strongSelf?.headerTitle.text = title
        strongSelf?.subTitle.text = subTitle
        strongSelf?.cancelButton.isHidden = true
        strongSelf?.okButton.addTarget(self, action: #selector(dismissView), for: .touchUpInside)
        okCompletion = nil
        cancelButton = nil
        self.animateToSuperView()
    }
    public func showAlertWithOkAction(_ title: String, subTitle: String, okCompletionHandler: @escaping OkCompletionHandler){
        let window: UIWindow = UIApplication.shared.keyWindow!
        window.addSubview(view)
        window.bringSubview(toFront: view)
        view.frame = window.bounds
        self.setUpContentView()
        //        view.addSubview(self.subView)
        strongSelf?.headerTitle.text = title
        strongSelf?.subTitle.text = subTitle
        strongSelf?.cancelButton.isHidden = true
        strongSelf?.okButton.addTarget(self, action: #selector(buttonActon(_:)), for: .touchUpInside)
        strongSelf?.okButton.tag = 1
        self.okCompletion = okCompletionHandler
        self.animateToSuperView()
    }
    public func showAlertWithTwoButtons(_ title: String, subTitle: String, okCompletionHandler : @escaping OkCompletionHandler, cancelCompletionHandler : @escaping CancelCompletionHandler){
        let window: UIWindow = UIApplication.shared.keyWindow!
        window.addSubview(view)
        window.bringSubview(toFront: view)
        view.frame = window.bounds
        self.setUpContentView()
        view.addSubview(contentView)
        strongSelf?.headerTitle.text = title
        strongSelf?.subTitle.text = subTitle
        strongSelf?.okButton.addTarget(self, action: #selector(buttonActon(_:)), for: .touchUpInside)
        strongSelf?.okButton.tag = 1
        strongSelf?.cancelButton.tag = 2
        strongSelf?.cancelButton.addTarget(self, action: #selector(buttonActon(_:)), for: .touchUpInside)
        strongSelf?.okCompletion = okCompletionHandler
        strongSelf?.cancelCompletion = cancelCompletionHandler
        self.animateToSuperView()
    }
    public func showAlertWithCustomButtons(_ title: String, subTitle: String, leftBtnTitle: String, leftBtnColor: UIColor, rightBtnTitle: String, rightBtnColor: UIColor,  leftCompletionHandler: @escaping CancelCompletionHandler , _ rightCompletionHandler: @escaping OkCompletionHandler){
        let window: UIWindow = UIApplication.shared.keyWindow!
        window.addSubview(view)
        window.bringSubview(toFront: view)
        view.frame = window.bounds
        self.setUpContentView()
        view.addSubview(contentView)
        strongSelf?.headerTitle.text = title
        strongSelf?.subTitle.text = subTitle
        strongSelf?.cancelButton.setTitle(leftBtnTitle, for: .normal)
        strongSelf?.cancelButton.backgroundColor = leftBtnColor
        strongSelf?.okButton.setTitle(rightBtnTitle, for: .normal)
        strongSelf?.okButton.backgroundColor = rightBtnColor
        
        strongSelf?.okButton.addTarget(self, action: #selector(buttonActon(_:)), for: .touchUpInside)
        strongSelf?.okButton.tag = 1
        strongSelf?.cancelButton.tag = 2
        strongSelf?.cancelButton.addTarget(self, action: #selector(buttonActon(_:)), for: .touchUpInside)
        strongSelf?.okCompletion = rightCompletionHandler
        strongSelf?.cancelCompletion = leftCompletionHandler
        self.animateToSuperView()
    }
    // MARK: - Animate View
    @objc func buttonActon(_ sender: UIButton) {
        switch sender.tag {
        case OkayButtonTag:
            if self.okCompletion != nil{
                self.okCompletion!()
            }
            break
        case CancelButtonTag:
            if self.cancelCompletion != nil{
                self.cancelCompletion!()
            }
            break
        default:
            break
        }
        self.dismissView()
        
    }
    @objc func dismissView() {
        
        UIView.animate(
            withDuration: 0.15,
            animations: {
            self.subView.transform = CGAffineTransform(scaleX: 0.20, y: 0.5)
        }) { _ in
            self.view.removeFromSuperview()
        }
    }
    fileprivate func animateToSuperView(){
        self.subView.transform = CGAffineTransform(scaleX: 0.25, y: 0.25 )
        UIView.animate(withDuration: 0.20, animations: {
            self.subView.transform = CGAffineTransform.identity
            
        })
    }

}


SKAlertTests/SKAlertTests.swift
//
//  SKAlertTests.swift
//  SKAlertTests
//
//  Created by Srujan k on 04/07/18.
//  Copyright © 2018 Srujan k. All rights reserved.
//

import XCTest
@testable import SKAlert

class SKAlertTests: XCTestCase {
    
    override func setUp() {
        super.setUp()
        // Put setup code here. This method is called before the invocation of each test method in the class.
    }
    
    override func tearDown() {
        // Put teardown code here. This method is called after the invocation of each test method in the class.
        super.tearDown()
    }
    
    func testExample() {
        // This is an example of a functional test case.
        // Use XCTAssert and related functions to verify your tests produce the correct results.
    }
    
    func testPerformanceExample() {
        // This is an example of a performance test case.
        self.measure {
            // Put the code you want to measure the time of here.
        }
    }
    
}


SKAlertExample/SKAlertExample/AppDelegate.swift
//
//  AppDelegate.swift
//  SKAlertExample
//
//  Created by Srujan k on 08/07/18.
//  Copyright © 2018 Srujan k. All rights reserved.
//

import UIKit
import CoreData

@UIApplicationMain
class AppDelegate: UIResponder, UIApplicationDelegate {

    var window: UIWindow?


    func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplicationLaunchOptionsKey: Any]?) -> Bool {
        // Override point for customization after application launch.
        return true
    }

    func applicationWillResignActive(_ application: UIApplication) {
        // Sent when the application is about to move from active to inactive state. This can occur for certain types of temporary interruptions (such as an incoming phone call or SMS message) or when the user quits the application and it begins the transition to the background state.
        // Use this method to pause ongoing tasks, disable timers, and invalidate graphics rendering callbacks. Games should use this method to pause the game.
    }

    func applicationDidEnterBackground(_ application: UIApplication) {
        // Use this method to release shared resources, save user data, invalidate timers, and store enough application state information to restore your application to its current state in case it is terminated later.
        // If your application supports background execution, this method is called instead of applicationWillTerminate: when the user quits.
    }

    func applicationWillEnterForeground(_ application: UIApplication) {
        // Called as part of the transition from the background to the active state; here you can undo many of the changes made on entering the background.
    }

    func applicationDidBecomeActive(_ application: UIApplication) {
        // Restart any tasks that were paused (or not yet started) while the application was inactive. If the application was previously in the background, optionally refresh the user interface.
    }

    func applicationWillTerminate(_ application: UIApplication) {
        // Called when the application is about to terminate. Save data if appropriate. See also applicationDidEnterBackground:.
        // Saves changes in the application's managed object context before the application terminates.
    }
}



SKAlertExample/SKAlertExample/SKAlertExampleViewController.swift
//
//  SKAlertExampleViewController.swift
//  SKAlertExample
//
//  Created by Srujan k on 08/07/18.
//  Copyright © 2018 Srujan k. All rights reserved.
//


///  OPTIONS:
/*
1.  If you pass title/ subtitle empty to SKAlert fucntion then that field automatically hides.
2. For alerts containig completion handlers. You will have to manaully set 'isTapToDismiss' variable to false, to disable tap anywhere to dismiss feature. */


import UIKit
import SKAlert

class SKAlertExampleViewController: UITableViewController {

    override func viewDidLoad() {
        super.viewDidLoad()
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
    }

    // MARK: - Table view data source
    override func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }

    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return 6
    }
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        switch indexPath.row {
        case 0:
            let alert = SKAlert()
            alert.isTapToDismiss = false
            alert.showAlert("Plain Alert Header.", subTitle: "Plain Alert Sub Title")
            break
        case 1:
            _  = SKAlert().showAlertWithOkAction("Alert Header", subTitle: "Alert Sub Title", okCompletionHandler: {
                print("Ok Pressed")
            })
            break
        case 2:
            _  = SKAlert().showAlertWithTwoButtons("Alert Header With Two Buttons", subTitle: "Alert Sub Title With Two Buttons", okCompletionHandler: {
                print("Ok Pressed")
            }, cancelCompletionHandler: {
                print("Canel Pressed")
            })
            break
        case 3:
            _ = SKAlert().showAlertWithCustomButtons("Alert Header With Two Custom Buttons", subTitle: "Sub title with custom buttons with Handlers........!!!", leftBtnTitle: "Left", leftBtnColor: UIColor.black, rightBtnTitle: "right", rightBtnColor: UIColor.red, leftCompletionHandler: {
                print("left")
            }) {
                print("right")
            }
            break
        case 4:
            _  = SKAlert().showAlert("Plain Alert Header with no subtitle", subTitle: "")
            break
        case 5:
            _  = SKAlert().showAlert("Header, Tap outside alert to dismiss..", subTitle: "")
            break
        default:
            break
        }
    }

}
